File: sections/100_grandmaster_finale/problems/63_periodic_forbidden_frontier/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;
const int MOD=1e9+7;
using Matrix=vector<vector<int>>;
Matrix multiply(const Matrix&a,const Matrix&b) {
    int n=a.size();
    Matrix c(n,vector<int>(n));
    for(int i=0;i<n;i++)for(int k=0;k<n;k++)if(a[i][k])for(int j=0;j<n;j++)c[i][j]=(c[i][j]+(ll)a[i][k]*b[k][j])%MOD;
    return c;
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    unsigned long long h;
    int p,w;
    cin>>h>>p>>w;
    vector<string>rows(p);
    for(auto&row:rows)cin>>row;
    int size=1<<w;
    Matrix block(size,vector<int>(size));
    for(int i=0;i<size;i++)block[i][i]=1;
    for(auto row:rows) {
        int blocked=0;
        for(int c=0;c<w;c++)if(row[c]=='#')blocked|=1<<c;
        Matrix transition(size,vector<int>(size));
        for(int incoming=0;incoming<size;incoming++)if(!(incoming&blocked)) {
            function<void(int,int,int)>fill=[&](int c,int used,int outgoing) {
                if(c==w) {
                    transition[incoming][outgoing]++;
                    return;
                }
                int bit=1<<c;
                if((used|blocked)&bit)fill(c+1,used,outgoing);
                else {
                    if(c+1<w&&!((used|blocked)&(bit<<1)))fill(c+2,used|bit|(bit<<1),outgoing);
                    fill(c+1,used|bit,outgoing|bit);
                }
            }
            ;
            fill(0,incoming,0);
        }
        block=multiply(block,transition);
    }
    Matrix answer(size,vector<int>(size));
    for(int i=0;i<size;i++)answer[i][i]=1;
    while(h) {
        if(h&1)answer=multiply(answer,block);
        block=multiply(block,block);
        h>>=1;
    }
    cout<<answer[0][0]<<'\n';
}

File: sections/100_grandmaster_finale/problems/63_periodic_forbidden_frontier/solution.py
import sys
MOD = 1000000007

def mul(a, b):
    n = len(a)
    c = [[0] * n for _ in range(n)]
    for i in range(n):
        for k, x in enumerate(a[i]):
            if x:
                for j, y in enumerate(b[k]):
                    c[i][j] = (c[i][j] + x * y) % MOD
    return c

def main():
    data = sys.stdin.buffer.read().split()
    h, p, w = map(int, data[:3])
    rows = [x.decode() for x in data[3:]]
    size = 1 << w
    block = [[int(i == j) for j in range(size)] for i in range(size)]
    for row in rows:
        blocked = sum(((row[c] == '#') << c for c in range(w)))
        mat = [[0] * size for _ in range(size)]
        for incoming in range(size):
            if incoming & blocked:
                continue

            def fill(c, used, out):
                if c == w:
                    mat[incoming][out] += 1
                    return
                bit = 1 << c
                if (used | blocked) & bit:
                    fill(c + 1, used, out)
                else:
                    if c + 1 < w and (not (used | blocked) & bit << 1):
                        fill(c + 2, used | bit | bit << 1, out)
                    fill(c + 1, used | bit, out | bit)
            fill(0, incoming, 0)
        block = mul(block, mat)
    result = [[int(i == j) for j in range(size)] for i in range(size)]
    while h:
        if h & 1:
            result = mul(result, block)
        block = mul(block, block)
        h //= 2
    print(result[0][0])
if __name__ == '__main__':
    main()
